### PR TITLE
[CI-IMPROVEMENT-6] Introduce targeted caching of tools

### PR DESCRIPTION
**Pull request by pavlovic-ivan**
#### What type of PR is this?
This is an improvement PR

#### What this PR does / why we need it:
Sixth in a series of PRs for improvements. This adds caching of tools from tools.yaml to the CI. This change cuts additional 2 minutes after cache is created. Iintegration tests will take less time, which will be the most visible improvement

#### Notes for maintainers
After PR is merged, i will add the same change with caching to workflows:

- codeql-analysis.yml
- airfow-operator.yml
- python-client.yml

### CODE REVIEW

File: magefiles/main.go
Comment (jgiannuzzi) [inline magefiles/main.go:21]: this type definition could be moved back inside `BootstrapTools` as it is not used anywhere else

File: magefiles/main.go
Comment (dave-gantenbein) [inline magefiles/main.go:26]: Should this and the following return err fmt.Printf the message? The standard in this file seems inconsistent, but I'd imagine I'd like to see a specific message if in fact something went wrong. Not a big deal, but you could be saving your future self some heartache. :)

File: magefiles/main.go
Comment (dave-gantenbein) [inline magefiles/main.go:275]: Is there a standard yaml package in this repo already we could use instead of hand rolling this?

File: magefiles/main.go
Comment (pavlovic-ivan) [reply magefiles/main.go:26]: returning an error here is actually raising an error to the main process that executed the function. that means that the main process will actually handle and log the error. does that answer your question?

File: magefiles/main.go
Comment (pavlovic-ivan) [reply magefiles/main.go:275]: i would assume there is, but i didn't look at that because of the approach used in the file, which is not using such library if exists. maybe it can be a future improvement, would you agree?

**Review by jgiannuzzi — APPROVED** [review]
LGTM 👍

File: magefiles/main.go
Comment (dave-gantenbein) [reply magefiles/main.go:275]: I'll defer to the rest of the team. :) I would not block merging for this.

Comment (richscott) [reply magefiles/main.go:275]: Theoretically, a `yaml.Encoder` object could be constructed (by opening the file, getting a `*File` handle (which implements the `io.Writer` interface), then call its `Encode(x interface{}) error` method, but it wouldn't be any more concise, because of all the boilerplate logic, e.g. (not tested):
```
func newWriteYAML(filename string, data interface{}) error {
	outfile, err := os.Open(filename)
	if err != nil {
		return err
	}
	encoder := yaml.NewEncoder(outfile)
	err = encoder.Encode(data)
	if err != nil {
		return err
	}
	err = encoder.Close()
	if err != nil {
		return err
	}
	return outfile.Close()
}
```
So I think the current code is about as brief & correct as any.

**Review by richscott — APPROVED** [review]
Looks great!

**Pull request merged**
